### ISSUE TITLE
How to implement Early-Stopping in MXnet?

### ISSUE THREAD

**Issue by wsstriving** [apache/incubator-mxnet#4965]
Hi,
     I want to do some scheduling in the training process:

1. halve the learning rate when the validation accuracy doesn't increase
2. stop the training when the learning rate is small enough

So I need to get the accuracy and learning rate history, but I can not find something like a monitor holding these information. Can anyone help?

**Comment by piiswrong** [comment]
You can add a batchend callback

**Comment by hantingxie** [comment]
Hello, @wsstriving @piiswrong. As I've seen quite a few issues for this topic, I'm planning to implement an EarlyStopping class in mxnet.callback (just like what Speedometer does) for stopping the training process for Python when the performance metric is not improving for a bit rounds (with a patience parameter for the control). That will be using the epoch_end_callback. 

It will need to modify both mxnet.callback.py and a little bit of base_module.py for the process to return.

Do you think this is necessary?

**Comment by piiswrong** [comment]
Its better to write the training loop yourself. Its actually pretty simple

**Comment by hantingxie** [comment]
Hello @piiswrong, thank you for the reply and your nice job for this framework. Sure, I agree that implementing a function for early stopping is not quite difficult.

However, my feeling is that early-stopping is a common feature in lots of other deep learning frameworks, and we even have that in R.  So this will be nice if we can have something similar that can make python users' life easier, too. Except for this, beginners who used other frameworks may find this more friendly. I've tried to make the new class multi-functional that can accept few useful parameters and should also be able to handle built-in metrics automatically.

Except for this, the other question we are facing now is that you can use callback functions for saving checkpoints of the model for early stopping, however, these callback functions cannot stop the fitting function without some minor modifications in base_module.py. This is also something that has also been covered in #6304 .

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
